akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ClusterShardingEntityProvider.scala

```diff
@@ -0,0 +1,141 @@
+        entity.stopMessage,
+        entity.entityProps,
+        clusterSettings,
+        shardingMessageExtractor,
+        None,
+        None,
+        None)
+
+    sharding.init(shardedEntity)
+  }
+
+  override def entityRefFor[M](typeKey: EntityTypeKey[M], entityId: String): EntityRef[M] =
+    toEntityRef(sharding.entityRefFor(toShardedEntityTypeKey(typeKey), entityId))
+
+}
+
+private object EntityAdapter {
+
+  def toShardedEntityTypeKey[M](typeKey: EntityTypeKey[M]): ShardedEntityTypeKey[M] =
+    typeKey match {
+      case shardedTypeKey: ShardedEntityTypeKey[M] @unchecked => shardedTypeKey
+      case typeKeyImpl: akka.actor.typed.EntityTypeKeyImpl[M] @unchecked =>
+        ShardedEntityTypeKeyImpl(typeKey.name, typeKeyImpl.messageClassName)
+    }
+
+  def toEntityContext[M](ctx: ShardedEntityContext[M]): EntityContext[M] = {
+    val typeKey = toEntityTypeKey(ctx.entityTypeKey)
+    // FIXME: needs proper conversion
+    val shardRef: ActorRef[EntityCommand] = ctx.shard.asInstanceOf[ActorRef[EntityCommand]]
+    new EntityContext(typeKey, ctx.entityId, shardRef)
+  }
+
+  def toEntityTypeKey[M](typeKey: ShardedEntityTypeKey[M]): EntityTypeKey[M] = {
+    val impl = typeKey.asInstanceOf[ShardedEntityTypeKeyImpl[M]]
+    akka.actor.typed.EntityTypeKeyImpl[M](typeKey.name, impl.messageClassName)
+  }
+
+  def toEntityRef[M](shardedEntity: ShardedEntityRef[M]): EntityRef[M] = new ShardedEntityRefWrapper(shardedEntity)
+
+  private[akka] class ShardedEntityRefWrapper[-M](delegate: ShardedEntityRef[M])
+      extends EntityRef[M]
+      with InternalRecipientRef[M] {
+
+    override def entityId: String = delegate.entityId
+
+    override def typeKey: EntityTypeKey[M] = toEntityTypeKey(delegate.typeKey)
+
+    override def tell(msg: M): Unit =
+      msg match {
+        case EntityEnvelope.StartEntity(entityId) =>
+          delegate.tell(ClassicStartEntity(entityId).asInstanceOf[M])
+        case entityEnv: EntityEnvelope[_] =>
+          val shardEnv = new ShardingEnvelope(entityEnv.entityId, entityEnv.message)
+          delegate.tell(shardEnv.asInstanceOf[M])
+        case _ => delegate.tell(msg)
+      }
+
+    override def ask[Res](f: ActorRef[Res] => M)(implicit timeout: Timeout): Future[Res] =
+      delegate.ask(f)(timeout)
+
+    override def askWithStatus[Res](f: ActorRef[StatusReply[Res]] => M)(implicit timeout: Timeout): Future[Res] =
+      delegate.askWithStatus(f)(timeout)
+
+    override def provider: ActorRefProvider =
+      delegate.asInstanceOf[InternalRecipientRef[M]].provider
+
+    override def isTerminated: Boolean =
+      delegate.asInstanceOf[InternalRecipientRef[M]].isTerminated
+
+    override def refPrefix: String =
+      URLEncoder.encode(s"${typeKey.name}-$entityId", ByteString.UTF_8)
+
+    override def toString: String = delegate.toString
+
+    override def hashCode(): Int = delegate.hashCode()
+    override def equals(obj: Any): Boolean = delegate.equals(obj)
```

Comment:
This two are probably not needed. Also not sure if a good idea to do full delegation.